### ISSUE TITLE
MicronautRequestHandler is not creating a Context - beans and properties do not inject

### ISSUE THREAD

**Issue by charlie-harvey** [micronaut-projects/micronaut-aws#732]
Thanks for reporting an issue, please review the task list below before submitting the
issue. Your issue report will be closed if the issue is incomplete and the below tasks not completed.

NOTE: If you are unsure about something and the issue is more of a question a better place to ask questions is on Stack Overflow (https://stackoverflow.com/tags/micronaut) or Gitter (https://gitter.im/micronautfw/). DO NOT use the issue tracker to ask questions.

### Task List

- [ x] Steps to reproduce provided
- [ x] Stacktrace (if present) provided
- [ x] Example that reproduces the problem uploaded to Github
- [ x] Full description of the issue provided (see below)

### Steps to Reproduce

1. Create function project: `mn create-function-app example.micronaut.func --features=aws-lambda,aws-v2-sdk --jdk=11 --lang=kotlin`
2. Add some value to `application.yml`
3. Try to get that value in your Handler using `@Value`

### Expected Behaviour

So anything I try to inject fails. I tried to `@Inject var lateinit s3Client: S3Client` and it fails. Tried a simple `@Value('stuff.thing')` and it also fails. What is frustrating is that it works in tests. So the `@MicronautTest` is creating an environment that has injection/introspection working. But having a `Handler` class in your sam_template.yml/serverless.yml fails.
Same process works just fine if I create an `app` instead of a `function`. So I created a repo with both a function and an app. Same `application.yml` file. Same `@Value()` annotation. Invoking the `app` in `sam local invoke` works great. Doing the `sam local invoke` in the `func` repo fails because it can't get the `@Value()`.

### Environment Information

- **Operating System**: Mac/AWS SAM/AWS Lambda (awslinux)
- **Micronaut Version:** 2.1.x
- **JDK Version:** 11

### Example Application

- https://github.com/charlie-harvey/micronaut-serverless-injection

**Comment by ttzn** [comment]
Add the following in `build.gradle`:

```
shadowJar {
    mergeServiceFiles()
}
```

This is actually required in **all projects**. Micronaut property loading uses the ServiceLoader API and reads `META-INF/services/io.micronaut.context.env.PropertySourceLoader` to find implementations, but many dependencies declare their own and they end up overriding each other. In your case it only contained:

```
io.micronaut.context.env.hocon.HoconPropertySourceLoader
```

With the fix it looks like:

```
io.micronaut.context.env.hocon.HoconPropertySourceLoader
io.micronaut.jackson.env.JsonPropertySourceLoader
io.micronaut.jackson.env.EnvJsonPropertySourceLoader
io.micronaut.jackson.env.CloudFoundryVcapApplicationPropertySourceLoader
io.micronaut.jackson.env.CloudFoundryVcapServicesPropertySourceLoader
io.micronaut.context.env.yaml.YamlPropertySourceLoader
io.micronaut.context.env.PropertiesPropertySourceLoader
```

Which is what you need to load all expected configuration files.

This issue has bitten so many people in the rear it needs to be dealt with. The `mergeServiceFiles()` option isn't mentioned in the main doc and the command line forgets to add it.

**Comment by graemerocher** [comment]
I believe it should be fixed in Micronaut 2.2 just released which automatically applies `mergeServiceFiles()` in the Gradle plugins

**Comment by charlie-harvey** [comment]
@ttzn Thanks so much for pointing that out. I feel less crazy now. Lost a few weeks pulling my hair out. @graemerocher Thanks for getting the fix in. I'll give it a try.
